File: src/skill_xray/scan.py
"""Run every detection over a parsed package and return unified findings.

The IR checks (taint over the OpenGrep lane, coverage, ... in ``checks``) read the parsed
structure and emit the shared Finding model; this module runs them and returns one
de-duplicated, severity-ordered list. When the operator passes a configured LLM client, an
advisory semantic pass is folded in on top of the deterministic findings. It reads only the IR."""

from __future__ import annotations

from copy import deepcopy
from dataclasses import asdict, dataclass, field

from .capability import build_triads
from .checks import run_checks
from .checks.coverage import is_inventory_note
from .correlate import correlate
from .disposition import apply_dispositions
from .findings import Finding, dedupe_findings
from .llm import adjudicate
from .llm.judge import POLICY_VERSION, REVIEW_POLICY_VERSION, judge_candidates
from .llm.session import LLMSession

__all__ = ["scan", "scan_report", "ScanReport"]


def _collect(parsed, executable, observations=None):
    return list(run_checks(parsed, opengrep_executable=executable,
                           **({"observations": observations} if observations is not None else {})))


def _advisory(parsed, client):
    try:
        return adjudicate(parsed, client)
    except Exception as exc:
        return [Finding(
            vector="", rule="llm-error", severity="low", path="",
            message="LLM adjudication pass failed (%s); deterministic findings stand"
                    % type(exc).__name__)]


def scan(parsed, *, client=None, opengrep_executable=None) -> list:
    """All findings for a parsed package, most severe first and deterministically ordered.

    Deterministic by default. When `client` is a configured LLM client (opt-in, the operator's
    own key), an advisory semantic-prompt-injection pass is added on top; it never removes or
    blocks a deterministic finding and fails closed on error."""
    findings = _collect(parsed, opengrep_executable)
    if client is not None:
        findings += _advisory(parsed, LLMSession(client))
    return dedupe_findings(findings)


@dataclass
class ScanReport:
    findings: list
    raw_candidates: list
    triads: dict
    shadow: list
    llm_usage: dict
    context_errors: list
    dispositions: list = field(default_factory=list)
    review_mode: bool = False
    correlation: dict = field(default_factory=dict)

    def to_dict(self):
        return deepcopy({"schema_version": "context-shadow-v1",
                         "findings": [finding.to_dict() for finding in self.findings],
                         "raw_scope": "emitted-check-results-before-reporting-deduplication",
                         "raw_candidates": self.raw_candidates,
                         "triads": {key: asdict(value) for key, value in self.triads.items()},
                         "shadow": self.shadow, "llm_usage": self.llm_usage,
                         "context_errors": self.context_errors,
                         "correlation": self.correlation,
                         **({"review_mode": "annotated", "dispositions": self.dispositions,
                             "final_findings": [f.to_dict() for f in self.findings]}
                            if self.review_mode else {})})


def scan_report(parsed, *, client=None, llm_shadow=False, opengrep_executable=None,
                max_llm_calls=25, llm_advisory=None, llm_review=False,
                disposition_policy=None) -> ScanReport:
    """Compatible opt-in context report. Candidate IDs are scan-local, not baseline identities."""
    if llm_shadow and llm_review:
        raise ValueError("Choose shadow or annotated LLM review, not both")
    review_enabled = llm_shadow or llm_review
    if review_enabled and client is None:
        raise ValueError("LLM review requires an explicitly supplied client")
    if llm_advisory is None:
        llm_advisory = not review_enabled
    session = LLMSession(client, max_calls=max_llm_calls) if client is not None else None
    observations = []
    raw = _collect(parsed, opengrep_executable, observations)
    gaps = {f.path for f in raw if not f.vector and not is_inventory_note(f.to_dict())}
    candidates = [
        {"candidate_id": "candidate-%06d" % i, "finding": deepcopy(f.to_dict()),
         "analyzer": f.evidence.get("engine", "ir-check"),
         "provenance": "deterministic-check-output",
         "coverage": "incomplete" if "" in gaps or f.path in gaps else "no-reported-gap"}
        for i, f in enumerate(raw)
    ]
    errors = []
    try:
        triads = build_triads(parsed, observations, raw)
    except Exception as exc:
        triads = {}
        errors.append("capability-context-error: %s" % type(exc).__name__)
    findings = list(raw)
    shadow = []
    dispositions = []
    if review_enabled:
        try:
            decisions = judge_candidates(parsed, candidates, triads, session,
                                         **({"apply_review": True} if llm_review else {}))
            if [d["candidate_id"] for d in decisions] != [c["candidate_id"] for c in candidates]:
                raise ValueError("Review candidate identity mismatch")
        except Exception as exc:
            errors.append("%s-review-error: %s" % (
                "llm" if llm_review else "shadow", type(exc).__name__))
            decisions = [{"candidate_id": c["candidate_id"], "disposition": "reported",
                          "status": "error", "proposal": None, "reason": "Review failed; retained",
                          "policy_version": REVIEW_POLICY_VERSION if llm_review else POLICY_VERSION,
                          "provenance": "deterministic-policy"} for c in candidates]
        if llm_review:
            dispositions = decisions
            # Model opinions must not change finding membership or severity.
        else:
            shadow = decisions
    supplemental = _advisory(parsed, session) if session is not None and llm_advisory else []
    findings += supplemental
    usage = session.usage() if session else {}
    if session is not None:
        usage.update(advisory_enabled=llm_advisory, judge_enabled=review_enabled)
    report_candidates = candidates + [
        {"candidate_id": "advisory-%06d" % i, "finding": deepcopy(f.to_dict()),
         "analyzer": "llm", "provenance": "advisory-output",
         "coverage": "no-reported-gap" if f.vector else "incomplete"}
        for i, f in enumerate(supplemental)]
    correlation = {"raw_candidates": deepcopy(report_candidates), "results": [], "links": []}
    try:
        correlation = correlate(parsed, report_candidates)
    except Exception as exc:
        errors.append("correlation-error: %s" % type(exc).__name__)
        correlation["errors"] = [errors[-1]]
    else:
        try:
            try:
                correlation = apply_dispositions(parsed, correlation, triads,
                                                 policy=disposition_policy, context_errors=errors)
            except ValueError as exc:
                errors.append("disposition-policy-error: %s" % type(exc).__name__)
                correlation = apply_dispositions(parsed, correlation, triads, context_errors=errors)
        except Exception as exc:
            errors.append("disposition-error: %s" % type(exc).__name__)
            correlation["errors"] = [errors[-1]]
    return ScanReport(dedupe_findings(findings), candidates, triads, shadow,
                      usage, errors, dispositions, llm_review, correlation)

File: internal/scan/scan.go
// Package scan runs every check over a parsed package (Python scan.py): Scan returns the
// deduplicated findings; Report also numbers the emitted candidates, builds the capability
// context, correlates and dispositions them and runs the opt-in LLM passes, each stage isolated
// so a failure is a visible context error and never a clean verdict.
package scan

import (
	"errors"
	"fmt"
	"slices"

	"github.com/traceforce/skill-xray/internal/capability"
	"github.com/traceforce/skill-xray/internal/checks"
	"github.com/traceforce/skill-xray/internal/correlate"
	"github.com/traceforce/skill-xray/internal/findings"
	"github.com/traceforce/skill-xray/internal/llm"
	"github.com/traceforce/skill-xray/internal/parse"
	"github.com/traceforce/skill-xray/internal/pytext"
)

// The names the oracle's tests monkeypatch on the scan module, as package vars.
var (
	RunChecks         = checks.Run
	adjudicate        = llm.Adjudicate
	BuildTriads       = capability.Build
	correlateFn       = correlate.Correlate
	applyDispositions = correlate.ApplyDispositions
	judgeCandidates   = llm.Judge
	// advisory is _advisory: the adjudication pass, whose failure becomes one low llm-error note
	// so the deterministic findings stand.
	advisory = func(p *parse.Package, s *llm.Session) (out []findings.Finding) {
		defer func() {
			if r := recover(); r != nil {
				out = []findings.Finding{{Rule: "llm-error", Severity: "low",
					Message: fmt.Sprintf("LLM adjudication pass failed (%T); deterministic findings stand", r)}}
			}
		}()
		return adjudicate(p, s, 25) // adjudicate._MAX_FILES
	}
)

// Scan is scan(): every deterministic finding plus, when a client is configured, the advisory
// LLM pass, deduplicated and severity-ordered.
func Scan(p *parse.Package, client llm.Completer, opengrepExe string) []findings.Finding {
	fs := RunChecks(p, opengrepExe, nil)
	if client != nil {
		s, _ := llm.NewSession(client, 25, 1<<20) // fixed non-negative budgets never fail
		fs = slices.Concat(fs, advisory(p, s))
	}
	return findings.Dedupe(fs)
}

// Options are scan_report's keyword arguments.
type Options struct {
	Client            llm.Completer
	LLMShadow         bool
	LLMReview         bool
	LLMApply          bool           // requires LLMReview
	LLMAdditive       bool           // run the advisory pass even under review (--llm-additive)
	OpengrepExe       string         // "" resolves the pinned binary
	DispositionPolicy map[string]any // nil is no policy
}

// ScanReport is the enrichment report; candidate ids are scan-local, not baseline identities.
type ScanReport struct {
	Findings      []findings.Finding
	RawCandidates []correlate.Candidate
	Triads        map[string]*capability.Triad
	Shadow        []llm.Decision
	LLMUsage      map[string]any
	ContextErrors []string
	Dispositions  []llm.Decision
	ReviewMode    bool
	Correlation   *correlate.Correlation
}

// ToMap is to_dict: the generic, deep-copied JSON view with integral numbers as int (D7).
func (r *ScanReport) ToMap() map[string]any {
	doc := map[string]any{"schema_version": "context-shadow-v1", "findings": r.Findings,
		"raw_scope":      "emitted-check-results-before-reporting-deduplication",
		"raw_candidates": r.RawCandidates, "triads": r.Triads, "shadow": r.Shadow,
		"llm_usage": r.LLMUsage, "context_errors": r.ContextErrors, "correlation": r.Correlation}
	if r.ReviewMode {
		doc["review_mode"], doc["dispositions"], doc["final_findings"] = "annotated", r.Dispositions, r.Findings
	}
	return pytext.JSONView(doc).(map[string]any)
}

// attempt runs fn and names its failure as Python names the exception class: the dynamic type
// of a recovered panic, or "ValueError" for a returned error (the correlate functions return
// only their policy and candidate-identity ValueErrors). "" is success.
func attempt(fn func() error) (name string) {
	defer func() {
		if r := recover(); r != nil {
			name = fmt.Sprintf("%T", r)
		}
	}()
	if fn() != nil {
		return "ValueError"
	}
	return ""
}

// Report is scan_report. The error is one of its three argument ValueErrors.
func Report(p *parse.Package, o Options) (*ScanReport, error) {
	if o.LLMShadow && o.LLMReview {
		return nil, errors.New("Choose shadow or annotated LLM review, not both")
	}
	if o.LLMApply && !o.LLMReview {
		return nil, errors.New("LLM-applied demotion requires annotated LLM review")
	}
	reviewEnabled := o.LLMShadow || o.LLMReview
	if reviewEnabled && o.Client == nil {
		return nil, errors.New("LLM review requires an explicitly supplied client")
	}
	llmAdvisory := o.LLMAdditive || !reviewEnabled
	var session *llm.Session
	if o.Client != nil {
		session, _ = llm.NewSession(o.Client, 25, 1<<20) // fixed non-negative budgets never fail
	}
	observations := []map[string]any{}
	raw := RunChecks(p, o.OpengrepExe, &observations)
	candidates := correlate.Candidates(raw)
	errs := []string{}
	triads := map[string]*capability.Triad{}
	if name := attempt(func() error { triads = BuildTriads(p, observations, raw); return nil }); name != "" {
		errs = append(errs, "capability-context-error: "+name)
	}
	shadow, dispositions := []llm.Decision{}, []llm.Decision{}
	if reviewEnabled {
		var decisions []llm.Decision
		name := attempt(func() error {
			decisions = judgeCandidates(p, candidates, triads, session, o.LLMReview)
			if len(decisions) != len(candidates) {
				return errors.New("Review candidate identity mismatch")
			}
			for i, d := range decisions {
				if d.CandidateID != candidates[i].CandidateID {
					return errors.New("Review candidate identity mismatch")
				}
			}
			return nil
		})
		if name != "" {
			kind, version := "shadow", llm.ShadowPolicyVersion
			if o.LLMReview {
				kind, version = "llm", llm.ReviewPolicyVersion
			}
			errs = append(errs, kind+"-review-error: "+name)
			decisions = make([]llm.Decision, len(candidates))
			for i, c := range candidates {
				decisions[i] = llm.Decision{CandidateID: c.CandidateID, Disposition: "reported", Status: "error",
					Reason: "Review failed; retained", PolicyVersion: version, Provenance: "deterministic-policy"}
			}
		}
		if o.LLMReview {
			dispositions = decisions // model opinions never change finding membership or severity
		} else {
			shadow = decisions
		}
	}
	supplemental := []findings.Finding{}
	if session != nil && llmAdvisory {
		supplemental = advisory(p, session)
	}
	usage := map[string]any{}
	if session != nil {
		usage = session.Usage()
		usage["advisory_enabled"], usage["judge_enabled"], usage["apply_enabled"] = llmAdvisory, reviewEnabled, o.LLMApply
	}
	reportCandidates := slices.Clone(candidates)
	for i, f := range supplemental {
		c := correlate.Candidate{CandidateID: fmt.Sprintf("advisory-%06d", i), Analyzer: "llm",
			Provenance: "advisory-output", Coverage: "no-reported-gap"}
		c.Finding = pytext.DeepCopy(f.ToMap()).(map[string]any)
		if f.Vector == "" {
			c.Coverage = "incomplete"
		}
		reportCandidates = append(reportCandidates, c)
	}
	var correlation *correlate.Correlation
	if name := attempt(func() (err error) { correlation, err = correlateFn(p, reportCandidates); return err }); name != "" {
		errs = append(errs, "correlation-error: "+name)
		correlation = &correlate.Correlation{RawCandidates: reportCandidates, Results: []*correlate.Result{},
			Links: []correlate.Link{}, Errors: []string{errs[len(errs)-1]}}
	} else {
		name := attempt(func() error {
			final, err := applyDispositions(p, correlation, triads, o.DispositionPolicy, errs)
			if err != nil {
				errs = append(errs, "disposition-policy-error: ValueError")
				if final, err = applyDispositions(p, correlation, triads, nil, errs); err != nil {
					return err
				}
			}
			correlation = final
			if o.LLMApply {
				reviews := make([]correlate.Review, len(dispositions))
				for i, d := range dispositions {
					reviews[i] = correlate.Review{CandidateID: d.CandidateID, Disposition: d.Disposition, Status: d.Status, Reason: d.Reason}
				}
				correlation = correlate.ApplyLLMReview(correlation, reviews)
			}
			return nil
		})
		if name != "" {
			errs = append(errs, "disposition-error: "+name)
			correlation.Errors = []string{errs[len(errs)-1]}
		}
	}
	return &ScanReport{Findings: findings.Dedupe(slices.Concat(raw, supplemental)), RawCandidates: candidates,
		Triads: triads, Shadow: shadow, LLMUsage: usage, ContextErrors: errs, Dispositions: dispositions,
		ReviewMode: o.LLMReview, Correlation: correlation}, nil
}
